src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -233,8 +233,71 @@ def requires_python38_or_newer(self) -> bool:
             allowed_versions=["3.8", "3.9", "3.10"], prior_version="3.7"
         )
 
+    def partition_by_major_minor_versions(self) -> tuple[InterpreterConstraints, ...]:
+        """Create a distinct InterpreterConstraints value for each CPython major-minor version that
+        is compatible with the original constraints."""
+        if any(req.project_name != "CPython" for req in self):
+            raise AssertionError(
+                "This function only works with CPython interpreter constraints for now."
+            )
+
+        def all_valid_patch_versions(major_minor: str, last_patch: int) -> list[int]:
+            return [
+                p
+                for p in range(0, last_patch + 1)
+                for req in self
+                if req.specifier.contains(f"{major_minor}.{p}")  # type: ignore[attr-defined]
+            ]
+
+        result = []
+
+        def maybe_add_version(major_minor: str, next_major_minor: str, last_patch: int) -> None:
+            valid_patch_versions = all_valid_patch_versions(major_minor, last_patch)
+            if not valid_patch_versions:
+                return
+
+            if len(valid_patch_versions) == 1:
+                result.append(
+                    InterpreterConstraints([f"=={major_minor}.{valid_patch_versions[0]}"])
+                )
+                return
+
+            min_constraint = (
+                f">={major_minor}"
+                if valid_patch_versions[0] == 0
+                else f">={major_minor}.{valid_patch_versions[0]}"
+            )
+            max_constraint = (
+                f"<{next_major_minor}"
+                if valid_patch_versions[-1] == last_patch
+                else f"<={major_minor}.{valid_patch_versions[-1]}"
+            )
+            merged_constraint = f"{min_constraint},{max_constraint}"
+
+            skipped = _not_in_contiguous_range(valid_patch_versions)
+            if skipped:
+                skipped_constraints = ",".join(f"!={major_minor}.{p}" for p in skipped)
+                merged_constraint += f",{skipped_constraints}"
+
+            result.append(InterpreterConstraints([merged_constraint]))
+
+        maybe_add_version("2.7", "2.8", 18)
+        maybe_add_version("3.5", "3.6", 11)
+        maybe_add_version("3.6", "3.7", 15)
+        maybe_add_version("3.7", "3.8", 15)
+        maybe_add_version("3.8", "3.9", 15)
+        maybe_add_version("3.9", "3.10", 15)
+        maybe_add_version("3.10", "3.11", 15)
```

Comment:
It really feels like it should not be necessary to include patch versions to do this. Whether or not a version `3.9.99` is valid shouldn't matter: if someone gives us `!=3.9.98`, it's always safe to include `>=3.9.99`, whether or not it actually exists. So can basically just increment the patch in that case.

On the other hand, including major,minor versions does seem necessary (to handle `>=3.5,<4`, etc), but it would be good for them not to be hardcoded... maybe the base set of "major,minor" versions that exist should be extendable via an option on `python-setup`?

Reply:
> It really feels like it should not be necessary to include patch versions to do this. Whether or not a version 3.9.99 is valid shouldn't matter: if someone gives us !=3.9.98, it's always safe to include >=3.9.99, whether or not it actually exists. So can basically just increment the patch in that case.


Patch versions do matter: we've seen that many users say things like "Do not run on 3.5.3", e.g. as a way to filter out system python.

> On the other hand, including major,minor versions does seem necessary (to handle >=3.5,<4, etc), but it would be good for them not to be hardcoded... maybe the base set of "major,minor" versions that exist should be extendable via an option on python-setup?

I think that could make sense? I do worry about the risk of someone trying to use Python 3.11 on Pants 2.7 a year from now. I don't like more options fatigue, but maybe that's less offensive.

Reply:
> > It really feels like it should not be necessary to include patch versions to do this. Whether or not a version 3.9.99 is valid shouldn't matter: if someone gives us !=3.9.98, it's always safe to include >=3.9.99, whether or not it actually exists. So can basically just increment the patch in that case.
> 
> Patch versions do matter: we've seen that many users say things like "Do not run on 3.5.3", e.g. as a way to filter out system python.

Yes, and I'm not saying they don't matter. I'm just saying that it's not necessary to hardcode their existence in this code, afaict.

Reply:
> I'm just saying that it's not necessary to hardcode their existence in this code, afaict.

Ah, I think it is. The only way I've found to robustly handle compatibility checks with Python requirement strings is to iterate over each patch version and check if it's contained or not. In the past, I tried doing things like analyzing version strings and could not get it to handle every edge case. We already have code like this:

https://github.com/pantsbuild/pants/blob/f0ca24624a43d2755f9fe6982341abfec13b4a45/src/python/pants/backend/python/util_rules/interpreter_constraints.py#L169-L185

We need some upper bound on when to stop iterating over patch versions. We could have it higher, but there's a perf hit from wasted work.

Do you know of another way to approach this? I agree it feels inelegant to iterate over every possible patch version.

Reply:
Re. the upper patch versions, it probably wouldn't hurt to keep the patch levels for actively maintained minor versions relatively high (e.g. 20), and then lock them down once the minor version has its last release. That'd be a relatively small performance trade-off. More generally, this doesn't sound like a particularly expensive process, particularly if you memoize the results later on.

Reply:
> Do you know of another way to approach this? I agree it feels inelegant to iterate over every possible patch version.

It looks like `Requirement.specifiers` is a [`SpecifierSet`](https://github.com/pypa/packaging/blob/d59efa2141210447ae21c4d48813d720e0d1dbfd/packaging/specifiers.py#L638), which can be iterated over to get [`_IndividualSpecifier`](https://github.com/pypa/packaging/blob/d59efa2141210447ae21c4d48813d720e0d1dbfd/packaging/specifiers.py#L97)s... yes, the latter type is private, but it's exposed out of a public type, soooo.

Rewriting a series of `_IndividualSpecifier`s (which separate the operator from the version) seems possible.

---

Failing that though, it looks like `SpecifierSet` also has a [`filter` method](https://github.com/pypa/packaging/blob/d59efa2141210447ae21c4d48813d720e0d1dbfd/packaging/specifiers.py#L777-L779), which might make doing a bunch of checks cheaper.

Reply:
> Rewriting a series of _IndividualSpecifiers (which separate the operator from the version) seems possible. 

And then trying to make semantic sense of what they all do? That feels more complex and error-prone than leveraging pkg_resources's `.contains()` method and exhaustive tests.

> which might make doing a bunch of checks cheaper.

Fwict, similar performance: under the hood, `.filters` is calling `.contains()` like we are already directly doing.